### DISCUSSION TITLE
External api for async inital state

### DISCUSSION THREAD

**Discussion by TorbjornHoltmon** [pmndrs/zustand#676] [Q&A] [answered]
Hi!

I want to try to use zustand for a project where the inital state is always fetched from an external api.
I have tried to figure out how to do this without any luck, have anyone done anything like this before and can give any pointers in the right direction?

```javascript
const useStore = create(set => ({
  // Fetch that gets the current bears from api
  bears: await getBearsAsync(),
  increasePopulation: () => set(state => ({ bears: state.bears + 1 })),
  removeAllBears: () => set({ bears: 0 })
}))
```

**Comment by dai-shi** [comment, marked as answer]
The initial value has to be something in sync.

```js
const useStore = create(set => ({
  bears: null,
  increasePopulation: () => set(state => ({ bears: state.bears + 1 })),
  removeAllBears: () => set({ bears: 0 })
}))

// Fetch that gets the current bears from api
getBearsAsync().then((bears) => useStore.setState({ bears }))
```

**Reply by kenneropia** [reply]
Hello @dai-shi, please sir would the “ getBearsAsync().then((bears) => useStore.setState({ bears }))” be best placed in a useEffect in the parent component?

**Reply by dai-shi** [reply]
It depends. It's initialization of the store, so place it where you initialize it.

`useEffect` would be good if you want to delay the execution, but it feels unlikely.

**Reply by dai-shi** [reply]
#2488 is related.

**Reply by kenneropia** [reply]
thanks for the response

**Comment by steaks** [comment]
I built a third-party library called [leo-query](https://leoquery.com/) that makes this easier. It handles gotchas with loading state, errors, caching, retries, stale data, etc.

```tsx
import {create} from "zustand";
import {query, hook} from "leo-query";

const useStore = create(set => ({
  // Fetch that gets the current bears from api
  bears: query(getBearsAsync),
  increasePopulation: () => set(state => ({ bears: state.bears + 1 })),
  removeAllBears: () => set({ bears: 0 })
}));

const useStoreAsync = hook(useStore, /*suspense*/ false);

const YourComponent = () => {
  const bears = useStoreAsync(s => s.bears);
  if (bears.isLoading) {
    return <div>Loading...</div>;
  }
  return <div>{bears.value}</div>;
};
```
